master/internal/api_experiment.go

```diff
@@ -0,0 +1,168 @@
+package internal
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+	"github.com/determined-ai/determined/proto/pkg/experimentv1"
+)
+
+var errExperimentNotFound = status.Error(codes.NotFound, "experiment not found")
+
+func canActivate(dbExp *model.Experiment) error {
+	if dbExp.State != model.PausedState {
+		return status.Error(codes.InvalidArgument, "experiment is not in a paused state")
+	}
+	return nil
+}
+
+func canPause(dbExp *model.Experiment) error {
+	if dbExp.State != model.ActiveState { // TODO does this make sense?
+		return status.Error(codes.InvalidArgument, "experiment is not in a active state")
+	}
+	return nil
+}
+
+func canCancel(dbExp *model.Experiment) error {
+	if _, exist := model.RunningStates[dbExp.State]; !exist {
+		return status.Error(codes.InvalidArgument, "experiment is not in a running state")
+	}
+	return nil
+}
+
+func canArchive(dbExp *model.Experiment) error {
```

Comment:
question: Should we make archive and unarchive idempotent?

Reply:
would we need to check whether the experiment is in a terminal state first? or we shouldn't care about that here and treat archive as just a toggle on the experiment thus not even checking the state of the experiment and avoiding potential race conditions?

Reply:
we chatted and we'd return whether 4xx if the experiment in terminal state or non existing